### ISSUE TITLE
Phase 6.1: Train Semantic Similarity Model (~20M params)

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#266]
## Parent Issue
Part of Epic #252: Enhanced AST-Aware Retriever Implementation

## Overview
Train a semantic similarity model to bridge gaps when deterministic matching fails.

## Model Architecture
```
Input: (query_ast, doc_ast) → Encoded representations → Similarity score

Encoder: Compositional AST Encoder
- Root embeddings: 64d (existing)
- Role embeddings: 16d (S/V/O/A)
- Structure encoding: TreeLSTM or GNN
- Output: 256d sentence embedding

Similarity: Cosine + learned calibration
```

## Training Data
Source: Semantic similarity pairs from corpus

### Positive pairs
- Adjacent sentences in same document
- Question-answer pairs from Q&A benchmark
- Paraphrases (if available)

### Negative pairs
- Random sentences from different documents
- Same-topic but different facts
- Hard negatives: shared roots but different predicates

### Data generation script
```python
def generate_training_pairs(corpus_path: Path, output_path: Path):
    """Generate training pairs for semantic similarity."""
    positive_pairs = []
    negative_pairs = []
    
    # Adjacent sentences (positive)
    for doc in load_corpus(corpus_path):
        for i in range(len(doc.sentences) - 1):
            positive_pairs.append((doc.sentences[i], doc.sentences[i+1], 1.0))
    
    # Random pairs (negative)
    all_sentences = [s for doc in corpus for s in doc.sentences]
    for _ in range(len(positive_pairs)):
        s1, s2 = random.sample(all_sentences, 2)
        negative_pairs.append((s1, s2, 0.0))
```

## Training Configuration
```yaml
model:
  embedding_dim: 256
  hidden_dim: 512
  num_layers: 2
  dropout: 0.1

training:
  batch_size: 64
  learning_rate: 1e-4
  epochs: 50
  early_stopping_patience: 5
  
loss: ContrastiveLoss (margin=0.5)
```

## Success Criteria
- [ ] Model trained and saved to models/semantic_similarity/
- [ ] Validation accuracy >85%
- [ ] No embedding collapse (mean cosine <0.5)
- [ ] Model size ~20M parameters
